src/svm/impls/binary/binary.c

```diff
@@ -491,142 +566,229 @@ PetscErrorCode SVMUpdate_Binary_Private(SVM svm)
+    /* http://www.lib.kobe-u.ac.jp/repository/90000225.pdf */
+    TRY( PetscObjectGetComm((PetscObject) H,&comm) );
+    TRY( MatDestroy(&svm_binary->J) );
+
+    if (p == 1) { /* Penalty type 1 */
+      TRY( SVMGetC(svm,&C) );
+
+      TRY( MatGetLocalSize(H,&m,&n) );
+      TRY( MatGetSize(H,&M,&N) );
+
+      TRY( MatCreateConstantDiagonal(comm,m,n,M,N,1. / C,&svm_binary->J) );
+      TRY( MatAssemblyBegin(svm_binary->J,MAT_FINAL_ASSEMBLY) );
+      TRY( MatAssemblyEnd(svm_binary->J,MAT_FINAL_ASSEMBLY) );
+    } else { /* Penalty type 2 */
+      TRY( SVMGetCp(svm,&Cp) );
+      TRY( SVMGetCn(svm,&Cn) );
+
+      TRY( VecDestroy(&svm_binary->diag) );
+      TRY( VecDuplicate(y,&svm_binary->diag) );
+      diag = svm_binary->diag;
+
+      TRY( VecGetSubVector(diag,svm_binary->is_p,&diag_p) );
+      TRY( VecSet(diag_p,1. / Cp) );
+      TRY( VecRestoreSubVector(diag,svm_binary->is_p,&diag_p) );
+
+      TRY( VecGetSubVector(diag,svm_binary->is_n,&diag_n) );
+      TRY( VecSet(diag_n,1. / Cn) );
+      TRY( VecRestoreSubVector(diag,svm_binary->is_n,&diag_n) );
+
+      TRY( MatCreateDiag(diag,&svm_binary->J) );
+    }
+
+    mats[0] = svm_binary->J;
+    mats[1] = H;
+    TRY( MatCreateSum(comm,2,mats,&HpJ) ); /* H = H + J */
+    TRY( MatDestroy(&H) );
+
+    H  = HpJ;
+  }
+
+  /* Set operator into QP formulation */
+  TRY( SVMGetQPS(svm,&qps) );
+  TRY( QPSGetQP(qps,&qp) );
+  TRY( QPSetOperator(qp,H) );
+
+  /* Decreasing reference counts */
+  TRY( MatDestroy(&X) );
+  TRY( MatDestroy(&H) );
+  PetscFunctionReturn(0);
+}
+
 #undef __FUNCT__
 #define __FUNCT__ "SVMSetUp_Binary"
 PetscErrorCode SVMSetUp_Binary(SVM svm)
 {
-  SVM_Binary *svm_binary = (SVM_Binary *) svm->data;
+  SVM_Binary  *svm_binary = (SVM_Binary *) svm->data;
 
-  QP          qp;
   QPS         qps;
+  QP          qp;
+
+  Mat         Xt;
+  Vec         y;
 
-  Vec         diag,diag_p,diag_n;
-  Mat         H,HpE,mats[2];
   Vec         e;
   Vec         x_init,x_init_p,x_init_n;
-
   Mat         Be = NULL;
   PetscReal   norm;
   Vec         lb;
-  Vec         ub,ub_p,ub_n;
-
-  Mat         Xt_training,X_training;
-  Vec         y;
-  PetscInt    n,m,N;
+  Vec         ub = NULL,ub_p,ub_n;
 
-  PetscReal   C,Cp,Cn;
   PetscInt    p;        /* penalty type */
   PetscInt    svm_mod;
   SVMLossType loss_type;
 
+  PetscReal   C,Cp,Cn;
+
+  PetscInt    i;
+
+  /* monitors */
   PetscBool   svm_monitor_set;
   void        *mctx;  /* monitor context */
 
   PetscFunctionBegin;
   if (svm->setupcalled) PetscFunctionReturn(0);
 
-  TRY( SVMGetLossType(svm,&loss_type) );
-  TRY( SVMGetMod(svm,&svm_mod) );
-  TRY( SVMGetPenaltyType(svm,&p) );
-
   if (svm->posttraincalled) {
     TRY( SVMUpdate_Binary_Private(svm) );
     PetscFunctionReturn(0);
   }
 
-  if (svm->hyperoptset) {
-    TRY( SVMGridSearch(svm) );
-  }
+  /* TODO generalize implementation of hyper parameter optimization */
+  if (svm->hyperoptset) { TRY( SVMGridSearch(svm) ); }
```

Comment:
Wrong code style (Developers guide 2.2.1 13,14)
```suggestion
  if (svm->hyperoptset) {
    TRY( SVMGridSearch(svm) );
  }
```

Reply:
Accepted.